data-products/src/main/scala/org/sunbird/analytics/exhaust/collection/BaseCollectionExhaustJob.scala

```diff
@@ -102,27 +101,51 @@ trait BaseCollectionExhaustJob extends BaseReportsJob with IJob with OnDemandExh
     val result: List[CollectionBatchResponse] = processBatches(userCachedDF, collectionBatches);
     result.foreach(f => JobLogger.log("Batch Status", Some(Map("status" -> f.status, "batchId" -> f.batchId, "executionTime" -> f.execTime, "message" -> f.statusMsg, "location" -> f.file)), INFO));
-    Metrics(totalRequests = Some(result.length), failedRequests = Some(result.count(x => x.status.toUpperCase() == "FAILED")), successRequests = Some(result.count(x => x.status.toUpperCase() == "SUCCESS")))
+    Metrics(totalRequests = Some(result.length), failedRequests = Some(result.count(x => x.status.toUpperCase() == "FAILED")), successRequests = Some(result.count(x => x.status.toUpperCase() == "SUCCESS")), Some(0))
   }
 
-  def executeOnDemand(custodianOrgId: String, userCachedDF: DataFrame)(implicit spark: SparkSession, fc: FrameworkContext, config: JobConfig): Metrics =  {
-    val modelParams = config.modelParams.getOrElse(Map[String, Option[AnyRef]]());
-    val storageConfig = getStorageConfig(config, "");
-    val requests = getRequests(jobId());
+  def executeOnDemand(custodianOrgId: String, userCachedDF: DataFrame)(implicit spark: SparkSession, fc: FrameworkContext, config: JobConfig): Metrics = {
+    val storageConfig = getStorageConfig(config, "")
+    val filteredRequests = filterRequests(getRequests(jobId()))
+    val requests: Array[JobRequest] = filteredRequests.getOrElse("unique", new Array[JobRequest](0))
     val totalRequests = new AtomicInteger(requests.length)
-    JobLogger.log("Total Requests are ", Some(Map("jobId" -> jobId(), "totalRequests" -> requests.length)), INFO)
-    val result = for (request <- requests) yield {
-      if (validateRequest(request)) {
-        val res = CommonUtil.time(processRequest(request, custodianOrgId, userCachedDF))
-        JobLogger.log("The Request is processed", Some(Map("requestId" -> request.request_id, "timeTaken" -> res._1, "remainingRequest" -> totalRequests.getAndDecrement())), INFO)
-        res._2
-      } else {
-        JobLogger.log("Invalid Request", Some(Map("requestId" -> request.request_id, "remainingRequest" -> totalRequests.getAndDecrement())), INFO)
-        markRequestAsFailed(request, "Invalid request")
+    JobLogger.log(s"Total distinct requests are: ${totalRequests.getAndDecrement()}", None, INFO)
+    val result: Array[JobRequest] = for (request <- requests) yield {
+      try {
+        if (validateRequest(request)) {
+          updateRequests(Array(request)) // Set the request status to PROCESSING when it picked up for processing
+          val res = CommonUtil.time(processRequest(request, custodianOrgId, userCachedDF))
+          JobLogger.log("The Request is processed", Some(Map("requestId" -> request.request_id, "timeTaken" -> res._1, "requestStatus" -> res._2.status, "remainingRequest" -> totalRequests.getAndDecrement())), INFO)
+          saveRequests(storageConfig, Array(res._2))
```

Comment:
Create a new method called saveRequest in OnDemandExhaustJob base class. 
This method should be executed in a future.
The zipping time should added to the total request time
The program should wait until all future calls are complete